proc-macros/src/new/render_client.rs

```diff
@@ -0,0 +1,134 @@
+use super::{RpcDescription, RpcMethod, RpcSubscription};
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+impl RpcDescription {
+	pub(super) fn render_client(&self) -> Result<TokenStream2, syn::Error> {
+		let jsonrpsee = self.jsonrpsee_client_path.as_ref().unwrap();
+
+		let trait_name = quote::format_ident!("{}Client", &self.trait_def.ident);
+
+		let super_trait = if self.subscriptions.is_empty() {
+			quote! { #jsonrpsee::traits::Client }
+		} else {
+			quote! { #jsonrpsee::traits::SubscriptionClient }
+		};
+
+		let method_impls =
+			self.methods.iter().map(|method| self.render_method(method)).collect::<Result<Vec<_>, _>>()?;
+		let sub_impls = self.subscriptions.iter().map(|sub| self.render_sub(sub)).collect::<Result<Vec<_>, _>>()?;
+
+		let async_trait = self.jrps_client_item(quote! { __reexports::async_trait });
+
+		// Doc-comment to be associated with the client.
+		let doc_comment = format!("Client implementation for the `{}` RPC API.", &self.trait_def.ident);
+
+		let trait_impl = quote! {
+			#[#async_trait]
+			#[doc = #doc_comment]
+			pub trait #trait_name: #super_trait {
+				#(#method_impls)*
+				#(#sub_impls)*
+			}
+
+			impl<T> #trait_name for T where T: #super_trait {}
+		};
+
+		Ok(trait_impl)
+	}
+
+	fn render_method(&self, method: &RpcMethod) -> Result<TokenStream2, syn::Error> {
+		// `jsonrpsee::Error`
+		let jrps_error = self.jrps_client_item(quote::format_ident!("Error"));
+		// Rust method to invoke (e.g. `self.<foo>(...)`).
+		let rust_method_name = &method.signature.sig.ident;
+		// List of inputs to put into `JsonRpcParams` (e.g. `self.foo(<12, "baz">)`).
+		// Includes `&self` receiver.
+		let rust_method_params = &method.signature.sig.inputs;
+		// Name of the RPC method (e.g. `foo_makeSpam`).
+		let rpc_method_name = self.rpc_identifier(&method.name);
+
+		// Called method is either `request` or `notification`.
+		// `returns` represent the return type of the *rust method* (`Result< <..>, jsonrpsee::Error`).
+		let (called_method, returns) = if let Some(returns) = &method.returns {
+			let called_method = quote::format_ident!("request");
+			let returns = quote! { Result<#returns, #jrps_error> };
+
+			(called_method, returns)
+		} else {
+			let called_method = quote::format_ident!("notification");
+			let returns = quote! { Result<(), #jrps_error> };
+
+			(called_method, returns)
+		};
+
+		// Encoded parameters for the request.
+		let parameters = if !method.params.is_empty() {
+			let serde_json = self.jrps_client_item(quote! { __reexports::serde_json });
+			let params = method.params.iter().map(|(param, _param_type)| {
+				quote! { #serde_json::to_value(&#param)? }
+			});
+
+			quote! {
+				vec![ #(#params),* ].into()
```

Comment:
so, this only supports parameters by position i.e, vec/array? 

I think that is alright because I think `params by name` is not used that often, maybe we could have an attribute for named params.

Currently, `named params` is the default in the proc macros but I think positional parameters is better as default.

Reply:
It's the client implementation, so since parameters are encoded under the hood I thought it's not that important to support encoding by name as well. If that's required, we sure can.